pkg/azurediscovery/discovery_azure.go

```diff
@@ -0,0 +1,111 @@
+package azurediscovery
+
+import (
+	"context"
+	"encoding/json"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2021-03-01/compute"
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2020-11-01/network"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+	"github.com/daspawnw/prometheus-aws-discovery/pkg/discovery"
+	log "github.com/sirupsen/logrus"
+)
+
+type DiscoveryClientAZURE struct {
+	TagPrefix    bool
+	Tag          string
+	Subscription string
+}
+
+func (d DiscoveryClientAZURE) GetInstances() ([]discovery.Instance, error) {
+	authorizer, err := auth.NewAuthorizerFromEnvironment()
+	if err != nil {
+		log.Error(err)
+		return nil, err
+	}
+
+	vmssinstanceClient := compute.NewVirtualMachineScaleSetVMsClient(d.Subscription)
+	vmssinstanceClient.Authorizer = authorizer
+	vmssClient := compute.NewVirtualMachineScaleSetsClient(d.Subscription)
+	vmssClient.Authorizer = authorizer
+	interfacesClient := network.NewInterfacesClient(d.Subscription)
+	interfacesClient.Authorizer = authorizer
+	interfaceIPConfigurationClient := network.NewInterfaceIPConfigurationsClient(d.Subscription)
+	interfaceIPConfigurationClient.Authorizer = authorizer
+	targetInstances := []discovery.Instance{}
+
+	for vmssList, err := vmssClient.ListAll(context.Background()); vmssList.NotDone(); err = vmssList.Next() {
+		if err != nil {
+			log.Error(err)
+			return nil, err
+		}
+
+		for _, vmss := range vmssList.Values() {
+			var metrics = []discovery.InstanceMetrics{}
+			if val, ok := vmss.Tags[d.Tag]; ok {
```

Comment:
Could we maybe already filter by tags?
https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2015-06-15/compute#VirtualMachineScaleSetVMsClient.List

Reply:
```filter - the filter to apply to the operation. Allowed values are 'startswith(instanceView/statuses/code, 'PowerState') eq true', 'properties/latestModelApplied eq true', 'properties/latestModelApplied eq false'.```

Doesn't look like there is a tag filter.

Reply:
https://docs.microsoft.com/de-de/azure/governance/resource-graph/samples/starter?tabs=azure-cli#list-resources-with-a-specific-tag-value ??

Reply:
Yeah ther seems to a way for filtering in the CLI. Though I'm not quite sure whether that translates to the Libraries above.. They explicitly talk about allowed values. Can do some research into this, but it would also require to introduce extra requests to first get all RGs, as List requires to pass in a RG as well (unlike ListAll).. That means extra requests for discovering RGs and another loop for discovering VMSS in the RGs.. From my perspective, that would probably be slower and won't introduce much in benefits..